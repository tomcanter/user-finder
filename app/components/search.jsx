var React =       require('react');
var Navigation =  require('react-router').Navigation;
var Reflux =      require('reflux');

var SearchForm =      require('./search-form.jsx');
var ResultsList =     require('./results.jsx');
var UserActions =     require('../actions/user');
var SearchUserStore = require('../stores/search-users');

module.exports = React.createClass({
  mixins: [Navigation, Reflux.ListenerMixin],

  getInitialState: function () {
    return {
      results: {
        items: []
      },
      query: ''
    };
  },

  handleSearchFormSubmit: function(event) {
    event.preventDefault();

    this.transitionTo('users', {
      username: this.refs.searchForm.getSearchTerm()
    });
  },

  search: function(username) {
    // Clear results before loading new set
    this.setState({
      results: {
        items: []
      },
      query: username
    });

    if (username) {
      UserActions.searchUsername(username);
    }
  },

  onResults: function(data) {
    this.setState(data);
  },

  componentWillReceiveProps: function(nextProps) {
    // Search if route param changes
    this.search(nextProps.params.username);
  },

  componentDidMount: function() {
    this.listenTo(SearchUserStore, this.onResults);

    // Search if URL params are present on page render
    this.search(this.props.params.username);
  },

  render: function() {
    return (
      <div className="Search">
        <h1 className="Search-title">Search for a user</h1>
        <SearchForm onUserSearch={this.handleSearchFormSubmit} query={this.state.query} ref="searchForm" />
        <ResultsList results={this.state.results} />
        <this.props.activeRouteHandler />
      </div>
    )
  }
});
