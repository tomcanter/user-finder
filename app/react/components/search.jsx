var React =       require('react');
var Router =      require('react-router');
var Navigation =  Router.Navigation;
var State =       Router.State;
var Reflux =      require('reflux');
var isEmpty =     require('lodash-node/modern/objects/isEmpty');

var SearchForm =      require('./search-form.jsx');
var Pagination =      require('./pagination.jsx');
var Results =         require('./results.jsx');
var ResultsMessage =  require('./results-message.jsx');
var UserActions =     require('actions/user');
var SearchUserStore = require('stores/search-users');

var Search = React.createClass({
  mixins: [Navigation, Reflux.ListenerMixin, State],

  getInitialState: function() {
    return {
      results: {
        items: []
      }
    };
  },

  handleSearchFormSubmit: function(event) {
    event.preventDefault();

    this.transitionTo('users', {}, {
      q: this.refs.searchForm.getSearchTerm(),
      page: 1
    });
  },

  search: function(query) {
    // Clear results before loading new set
    this.setState({
      results: {
        items: []
      },
      query: query
    });

    if (!isEmpty(query)) {
      UserActions.searchUser(query);
    }
  },

  onResults: function(data) {
    this.setState(data);
  },

  componentWillReceiveProps: function() {
    // Search if route param changes
    this.search(this.getQuery());
  },

  componentDidMount: function() {
    this.listenTo(SearchUserStore, this.onResults);

    // Search if URL params are present on page render
    this.search(this.getQuery());
  },

  render: function() {
    return (
      <div className="Search">
        <div className="Search-item Search-wrapSearchForm">
          <div className="Container">
            <SearchForm onUserSearch={this.handleSearchFormSubmit} query={this.state.query} ref="searchForm" />
            <ResultsMessage results={this.state.results} query={this.state.query} />
          </div>
        </div>
        <div className="Container">
          <div className="Search-item">
            <Results results={this.state.results} />
          </div>
          <div className="Search-item u-mXA u-size8of10 u-md-size5of10">
            <Pagination results={this.state.results} perpage="30" />
          </div>
        </div>
      </div>
    )
  }
});

module.exports = Search;
