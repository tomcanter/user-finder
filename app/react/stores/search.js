import Reflux from 'reflux';
import User from 'actions/user';
import pick from 'lodash-node/modern/objects/pick';
import map from 'lodash-node/modern/collections/map';
import { setItem } from 'mixins/cache';
import parseLinkHeader from 'utils/parseLinkHeader';

export default Reflux.createStore({
  init() {
    this.listenTo(User.search.completed, this.onSearchCompleted);
    this.listenTo(User.search.failed, this.onSearchFailed);
  },

  onSearchCompleted(response, fromCache) {
    if (!fromCache) {
      // Only take data that's needed
      let results = pick(response.data, 'items', 'total_count');
      results.items = map(response.data.items, item => pick(item, 'avatar_url', 'id', 'login'));

      // Construct data to cache
      let data = {
        results,
        url: response.url,
        pagination: parseLinkHeader(response.xhr.getResponseHeader('Link') || '')
      };

      // Cache it
      setItem(data.url, data);

      // Notify the views
      this.trigger(data);
    } else {
      // Data came from cache, so send straight to views
      this.trigger(response);
    }
  },

  onSearchFailed(query, xhr) {
    this.trigger({
      results: {
        items: [],
        error: JSON.parse(xhr.responseText)
      },
      query
    });
  }
});
