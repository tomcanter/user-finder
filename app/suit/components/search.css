/** @define Search; use strict */

@import "suitcss-utils-size";
@import "./results.css";
@import "./searchform.css";
@import "./results-message.css";
@import "./pagination.css";
@import "../utils/space.css";

.Search-item {
  margin-bottom: 1.5rem;
}

.Search-item:last-child {
  margin-bottom: 0;
}

.Search-wrapSearchForm {
  border-bottom: 1px solid #ddd;
}