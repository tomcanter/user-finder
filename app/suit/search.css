/** @define Search; use strict */

@import "./searchform.css";
@import "./results-message.css";

.Search-item {
  margin-bottom: 1.5rem;
}

.Search-item:last-child {
  margin-bottom: 0;
}

.Search-item--bordered {
  border-bottom: 1px solid #ddd;
}