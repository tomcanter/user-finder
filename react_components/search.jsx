var React =       require('react');
var req =         require('reqwest');
var SearchForm =  require('./search-form.jsx');
var ResultsList = require('./results-list.jsx');
var Navigation =  require('react-router').Navigation;

module.exports = React.createClass({
  mixins: [Navigation],
  getInitialState: function () {
    return { results: { items: [] }}
  },
  handleSearchFormSubmit: function(event) {
    event.preventDefault();
    this.transitionTo('search', { username: this.refs.searchForm.getSearchTerm() })
  },
  getUsers: function(query) {
    this.setState({results: { items: [] }, query: query})

    if (!query) {
      return
    }

    req({
      url: 'https://api.github.com/search/users',
      data: {q: query},
      type: 'json'
    }).then(function(data) {
      this.setState({results: data, query: query})
    }.bind(this))
  },
  componentWillReceiveProps: function(nextProps) {
    this.getUsers(nextProps.params.username);
  },
  componentDidMount: function() {
    this.getUsers(this.props.params.username);
  },
  render: function() {
    return (
      <div className="Search">
        <h1 className="Search-title">Search for a user</h1>
        <SearchForm onUserSearch={this.handleSearchFormSubmit} query={this.state.query} ref="searchForm" />
        <ResultsList results={this.state.results} />
      </div>
    )
  }
});
