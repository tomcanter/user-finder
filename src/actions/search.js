// @flow

import pick from 'lodash/fp/pick';
import flow from 'lodash/fp/flow';
import get from 'lodash/fp/get';
import mapValues from 'lodash/fp/mapValues';

const pickUserData = flow(
  get('entities.users'),
  mapValues(
    pick([
      'login',
      'id',
      'avatar_url',
    ])
  )
);

export const SEARCH_REQUEST: string = 'SEARCH_REQUEST';
export function searchRequest({searchTerm}: {searchTerm: string}) {
  return {
    type: SEARCH_REQUEST,
    searchTerm,
  };
}

export const SEARCH_SUCCESS: string = 'SEARCH_SUCCESS';
export function searchSuccess(data: Object) {
  const {result, entities} = data;
  return {
    type: SEARCH_SUCCESS,
    users: result,
    entities,
    totalResults: result.length,
  };
}

export const SEARCH_FAILURE: string = 'SEARCH_FAILURE';
export function searchFailure(error: Object) {
  const {response, message} = error;
  return {
    type: SEARCH_FAILURE,
    response,
    message,
  };
}
