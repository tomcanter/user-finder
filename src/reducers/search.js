import pick from 'lodash/fp/pick';
import mergeAll from 'lodash/fp/mergeAll';
import {
  SEARCH_REQUEST,
  SEARCH_SUCCESS,
  SEARCH_FAILURE,
} from '../actions/search';

const initialState = {
  error: null,
  isPending: false,
  query: '',
  result: [],
  totalResults: null,
};

export default function searchReducer(state = initialState, action) {
  switch (action.type) {
    case SEARCH_REQUEST:
      return {
        ...state,
        isPending: true,
      };
    case SEARCH_SUCCESS:
      return mergeAll([
        state,
        pick(['result', 'query', 'totalResults'], action),
        {isPending: false},
      ]);
    case SEARCH_FAILURE: {
      const {response, message} = action;
      return {
        ...state,
        isPending: false,
        error: {
          response,
          message,
        },
      };
    }
    default:
      return state;
  }
}
