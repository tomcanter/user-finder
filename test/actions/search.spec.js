import * as actions from '../../src/actions/search';

describe('when requesting users via a search term', () => {
  it('should dispatch a SEARCH_REQUEST action', () => {
    expect(
      actions.searchRequest({searchTerm: 'tom'})
    ).toMatchSnapshot();
  });
});

describe('when receiving a set of users', () => {
  it('should dispatch a SEARCH_SUCCESS action with the necessary data', () => {
    const data = {
      entities: {
        users: {
          123: {login: 'foo', id: 123, other: 'test'},
          456: {login: 'baz', id: 456, other: 'test'},
        },
      },
      result: [123, 456],
    };
    expect(
      actions.searchSuccess(data)
    ).toMatchSnapshot();
  });
});

describe('when an error occurs during a request', () => {
  it('should dispatch a SEARCH_FAILURE action', () => {
    const error = new Error('something broke');
    error.response = {
      statusText: '404',
    };
    expect(
      actions.searchFailure(error)
    ).toMatchSnapshot();
  });
});
