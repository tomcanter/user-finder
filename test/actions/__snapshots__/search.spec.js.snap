exports[`when an error occurs during a request should dispatch a SEARCH_FAILURE action 1`] = `
Object {
  "message": "something broke",
  "response": Object {
    "statusText": "404",
  },
  "type": "SEARCH_FAILURE",
}
`;

exports[`when receiving a set of users should dispatch a SEARCH_SUCCESS action with the necessary data 1`] = `
Object {
  "entities": Object {
    "users": Object {
      "123": Object {
        "id": 123,
        "login": "foo",
        "other": "test",
      },
      "456": Object {
        "id": 456,
        "login": "baz",
        "other": "test",
      },
    },
  },
  "totalResults": 2,
  "type": "SEARCH_SUCCESS",
  "users": Array [
    123,
    456,
  ],
}
`;

exports[`when requesting users via a search term should dispatch a SEARCH_REQUEST action 1`] = `
Object {
  "searchTerm": "tom",
  "type": "SEARCH_REQUEST",
}
`;
